File: cpp/questions/productExceptSelf.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        std::vector<int> ans(n);    
        int pp = 1;
        for (int i=0; i<n; ++i) {
            ans[i] = pp;
            pp *= nums[i];
        }
        pp = 1;
        for (int i=n-1; i>=0; --i) {
            ans[i] *= pp;
            pp *= nums[i];
        }
        return ans;
    }
};

File: rust/questions/product_except_self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let nl = nums.len();
        let mut ans = vec![1; nl];
        for p in (0..nl - 1).rev() {
            ans[p] = nums[p + 1] * ans[p + 1];
        }
        let mut pre = nums[0];
        for p in (1..nl) {
            ans[p] *= pre;
            pre *= nums[p];
        }
        ans
    }
}
